farmerbot/internal/find_node.go

```diff
@@ -0,0 +1,154 @@
+package internal
+
+import (
+	"fmt"
+	"slices"
+	"sort"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	zos "github.com/threefoldtech/zos/client"
+)
+
+// FindNode finds an available node in the farm
+func (f *FarmerBot) findNode(sub Sub, nodeOptions NodeOptions) (uint32, error) {
+	log.Info().Msg("Finding a node")
+
+	nodeOptionsCapacity := capacity{
+		hru: nodeOptions.HRU,
+		sru: nodeOptions.SRU,
+		cru: nodeOptions.CRU,
+		mru: nodeOptions.MRU,
+	}
+
+	if (len(nodeOptions.GPUVendors) > 0 || len(nodeOptions.GPUDevices) > 0) && nodeOptions.HasGPUs == 0 {
+		// at least one gpu in case the user didn't provide the amount
+		nodeOptions.HasGPUs = 1
+	}
+
+	log.Debug().Interface("required filter options", nodeOptions)
+
+	if nodeOptions.PublicIPs > 0 {
+		var publicIpsUsedByNodes uint64
+		for _, node := range f.nodes {
+			publicIpsUsedByNodes += node.publicIPsUsed
+		}
+
+		if publicIpsUsedByNodes+nodeOptions.PublicIPs > uint64(len(f.farm.PublicIPs)) {
+			return 0, fmt.Errorf("not enough public ips available for farm %d", f.farm.ID)
+		}
+	}
+
+	var possibleNodes []node
+	for _, node := range f.nodes {
+		gpus := node.gpus
+		if nodeOptions.HasGPUs > 0 {
+			if len(nodeOptions.GPUVendors) > 0 {
+				gpus = filterGPUs(gpus, nodeOptions.GPUVendors, false)
+			}
+
+			if len(nodeOptions.GPUDevices) > 0 {
+				gpus = filterGPUs(gpus, nodeOptions.GPUDevices, true)
+			}
+
+			if len(gpus) < int(nodeOptions.HasGPUs) {
+				continue
+			}
+		}
+
+		if nodeOptions.Certified && !node.Certification.IsCertified {
+			continue
+		}
+
+		if nodeOptions.PublicConfig && !node.PublicConfig.HasValue {
+			continue
+		}
+
+		if node.hasActiveRentContract {
+			continue
+		}
+
+		if nodeOptions.Dedicated {
+			if !node.dedicated || !node.isUnused() {
+				continue
+			}
+		} else {
+			if node.dedicated && nodeOptionsCapacity != node.resources.total {
+				continue
+			}
+		}
+
+		if slices.Contains(nodeOptions.NodeExclude, uint32(node.ID)) {
+			continue
+		}
+
+		if !node.canClaimResources(nodeOptionsCapacity, f.config.Power.OverProvisionCPU) {
+			continue
+		}
+
+		possibleNodes = append(possibleNodes, node)
+	}
+
+	if len(possibleNodes) == 0 {
+		return 0, fmt.Errorf("could not find a suitable node with the given options: %+v", nodeOptions)
+	}
+
+	// Sort the nodes on power state (the ones that are ON first then waking up, off, shutting down)
+	sort.Slice(possibleNodes, func(i, j int) bool {
+		return possibleNodes[i].ID < possibleNodes[j].ID
+	})
+
+	sort.Slice(possibleNodes, func(i, j int) bool {
+		return possibleNodes[i].powerState < possibleNodes[j].powerState
+	})
```

Comment:
you can do the sorting in one call 

return ith powerstate < jth powerstate || ith.ID < jth.ID I think